File: beam_eye_tracker_sdk-2.1.0/cpp/samples/data_access_methods/bet_sync_data_access.cpp
/**
 * Copyright (c) 2025, Eyeware Tech SA.
 *
 * All rights reserved.
 *
 * @brief This sample demonstrates how to use the synchronous data access method.
 */
#include "../bet_sample_utils.h"
#include "eyeware/beam_eye_tracker.h"

void main() {
    eyeware::beam_eye_tracker::API bet_api("Sync Sample",
                                           eyeware::beam_eye_tracker::ViewportGeometry());

    // We hold a timestamp instance useful to sync the data reception.
    eyeware::beam_eye_tracker::Timestamp last_update_timestamp_sec{
        eyeware::beam_eye_tracker::NULL_DATA_TIMESTAMP};

    std::unique_ptr<eyeware::beam_eye_tracker::TrackingStateSet> last_received_tracking_state_set;

    eyeware::beam_eye_tracker::TrackingDataReceptionStatus previous_status =
        bet_api.get_tracking_data_reception_status();

    print_tracking_data_reception_status(previous_status);

    int count = 0;
    // Access 1 minute of data (assuming a 30fps webcam)
    while (count < 30 * 60) {
        const unsigned int wait_timeout_ms = 1000;

        // Follow up with the status of the tracking data reception.
        eyeware::beam_eye_tracker::TrackingDataReceptionStatus status =
            bet_api.get_tracking_data_reception_status();
        print_tracking_data_reception_status_if_changed(previous_status, status);
        previous_status = status;

        // Wait for a new frame. If this function returns true, then there is a new frame and
        // last_update_timestamp_sec is updated.
        if (bet_api.wait_for_new_tracking_state_set(last_update_timestamp_sec, wait_timeout_ms)) {
            // This is how we access the latest TrackingStateSet.
            // It is only movable, thus this is a way to retain it.
            last_received_tracking_state_set =
                std::make_unique<eyeware::beam_eye_tracker::TrackingStateSet>(
                    std::move(bet_api.get_latest_tracking_state_set()));

            print_latest_tracking_state_set(*last_received_tracking_state_set,
                                            last_update_timestamp_sec);
            count++;
        }
    }
}

File: beam_eye_tracker_sdk-2.1.0/python/samples/data_access_methods/bet_sync_data_access.py
# Copyright (c) 2025, Eyeware Tech SA.
#
# All rights reserved.
#
# This sample demonstrates how to use the synchronous data access method.
#
# Synchronous data access is convenient in case you want to get a TrackingStateSet at low
# latency, and it is fine to block the thread while waiting for data updates.
#
# A typical scenario is when your application (or at least this thread) uses the tracking data
# updates as the event to trigger further processing/recording/interactions, etc.
import sys
import os

sys.path.append(os.path.join(os.path.dirname(__file__), ".."))

import bet_add_to_python_path

from eyeware import beam_eye_tracker
from bet_sample_utils import (
    print_tracking_data_reception_status,
    print_latest_tracking_state_set,
    print_tracking_data_reception_status_if_changed,
)


def main():
    # Initialize API with empty viewport geometry
    viewport_geometry = beam_eye_tracker.ViewportGeometry()
    viewport_geometry.point_00 = beam_eye_tracker.Point(0, 0)
    viewport_geometry.point_11 = beam_eye_tracker.Point(0, 0)

    bet_api = beam_eye_tracker.API("Python Sync Sample", viewport_geometry)

    # We hold a timestamp instance useful to sync the data reception
    last_update_timestamp_sec = beam_eye_tracker.NULL_DATA_TIMESTAMP()

    last_received_tracking_state_set = None
    previous_status = bet_api.get_tracking_data_reception_status()
    print_tracking_data_reception_status(previous_status)

    count = 0
    # Access 1 minute of data (assuming a 30fps webcam)
    while count < 30 * 60:
        wait_timeout_ms = 1000

        # Follow up with the status of the tracking data reception
        status = bet_api.get_tracking_data_reception_status()
        print_tracking_data_reception_status_if_changed(previous_status, status)
        previous_status = status

        # Wait for a new frame. If this function returns true, then there is a new frame and
        # last_update_timestamp_sec is updated
        if bet_api.wait_for_new_tracking_state_set(last_update_timestamp_sec, wait_timeout_ms):
            # This is how we access the latest TrackingStateSet
            last_received_tracking_state_set = bet_api.get_latest_tracking_state_set()
            print_latest_tracking_state_set(last_received_tracking_state_set, last_update_timestamp_sec)
            count += 1


if __name__ == "__main__":
    main()
